File: lld/problems/tic-tac-toe/go/tictactoe.go
// Package tictactoe implements the classic Tic-Tac-Toe LLD problem on a
// generic NxN board: move validation, alternating turns between N players,
// and win detection via incrementally maintained row/column/diagonal counts.
package tictactoe

import (
	"errors"
	"fmt"
)

type Symbol string

const Empty Symbol = ""

type Player struct {
	Name   string
	Symbol Symbol
}

type GameStatus int

const (
	InProgress GameStatus = iota
	Won
	Draw
)

var (
	ErrOutOfBounds  = errors.New("move is out of bounds")
	ErrCellOccupied = errors.New("cell is already occupied")
	ErrGameOver     = errors.New("game is already over")
)

// Board is a generic NxN grid. Win detection is O(1) per move: instead of
// rescanning the whole board, we keep running counts per row, column and
// the two diagonals for whichever symbol is placed. A move only ever
// affects one row, one column and (at most) two diagonals, so checking
// those counts after each move is enough to detect a win.
type Board struct {
	size        int
	cells       [][]Symbol
	rowCounts   map[Symbol][]int
	colCounts   map[Symbol][]int
	diagCount   map[Symbol]int
	antiCount   map[Symbol]int
	filledCells int
}

func NewBoard(size int) *Board {
	cells := make([][]Symbol, size)
	for i := range cells {
		cells[i] = make([]Symbol, size)
	}
	return &Board{
		size:      size,
		cells:     cells,
		rowCounts: make(map[Symbol][]int),
		colCounts: make(map[Symbol][]int),
		diagCount: make(map[Symbol]int),
		antiCount: make(map[Symbol]int),
	}
}

func (b *Board) Size() int { return b.size }

func (b *Board) At(row, col int) Symbol { return b.cells[row][col] }

func (b *Board) inBounds(row, col int) bool {
	return row >= 0 && row < b.size && col >= 0 && col < b.size
}

func (b *Board) IsFull() bool { return b.filledCells == b.size*b.size }

// place records sym at (row, col) and returns true if that move completes
// a row, column, or diagonal of length size for sym.
func (b *Board) place(row, col int, sym Symbol) (bool, error) {
	if !b.inBounds(row, col) {
		return false, ErrOutOfBounds
	}
	if b.cells[row][col] != Empty {
		return false, ErrCellOccupied
	}
	b.cells[row][col] = sym
	b.filledCells++

	if _, ok := b.rowCounts[sym]; !ok {
		b.rowCounts[sym] = make([]int, b.size)
		b.colCounts[sym] = make([]int, b.size)
	}
	b.rowCounts[sym][row]++
	b.colCounts[sym][col]++
	if row == col {
		b.diagCount[sym]++
	}
	if row+col == b.size-1 {
		b.antiCount[sym]++
	}

	won := b.rowCounts[sym][row] == b.size ||
		b.colCounts[sym][col] == b.size ||
		b.diagCount[sym] == b.size ||
		b.antiCount[sym] == b.size
	return won, nil
}

func (b *Board) String() string {
	s := ""
	for r := 0; r < b.size; r++ {
		for c := 0; c < b.size; c++ {
			ch := b.cells[r][c]
			if ch == Empty {
				ch = "."
			}
			s += fmt.Sprintf("%s ", ch)
		}
		s += "\n"
	}
	return s
}

// Game orchestrates turn order, move validation and terminal-state tracking
// on top of a Board. It has no locking: unlike the parking lot, a single
// tic-tac-toe game is played by one turn owner at a time, so there is no
// concurrent-writer scenario to guard against.
type Game struct {
	board   *Board
	players []*Player
	turn    int
	status  GameStatus
	winner  *Player
}

func NewGame(size int, players []*Player) *Game {
	return &Game{board: NewBoard(size), players: players}
}

func (g *Game) Board() *Board { return g.board }

func (g *Game) Status() GameStatus { return g.status }

func (g *Game) Winner() *Player { return g.winner }

func (g *Game) CurrentPlayer() *Player { return g.players[g.turn] }

// Move plays a cell for the current player, advances the turn, and updates
// game status. Returns the player who just moved won and any validation error.
func (g *Game) Move(row, col int) (bool, error) {
	if g.status != InProgress {
		return false, ErrGameOver
	}

	player := g.players[g.turn]
	won, err := g.board.place(row, col, player.Symbol)
	if err != nil {
		return false, err
	}

	if won {
		g.status = Won
		g.winner = player
		return true, nil
	}
	if g.board.IsFull() {
		g.status = Draw
		return false, nil
	}

	g.turn = (g.turn + 1) % len(g.players)
	return false, nil
}

File: lld/problems/tic-tac-toe/python/tic_tac_toe.py
"""Tic-Tac-Toe LLD — Python reference implementation.

Generic NxN board, N alternating players, incremental win detection via
row/column/diagonal counters. See ../README.md for the design writeup.
"""
from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field
from enum import Enum, auto
from typing import List, Optional


class GameStatus(Enum):
    IN_PROGRESS = auto()
    WON = auto()
    DRAW = auto()


@dataclass
class Player:
    name: str
    symbol: str


class OutOfBoundsError(Exception):
    pass


class CellOccupiedError(Exception):
    pass


class GameOverError(Exception):
    pass


class Board:
    """A generic NxN grid.

    Win detection is O(1) per move: instead of rescanning the whole board,
    we keep running counts per row, column and the two diagonals for
    whichever symbol is placed. A move only ever affects one row, one
    column and (at most) two diagonals, so checking those counts after
    each move is enough to detect a win.
    """

    def __init__(self, size: int = 3):
        self.size = size
        self._cells: List[List[Optional[str]]] = [[None] * size for _ in range(size)]
        self._row_counts = defaultdict(lambda: [0] * size)
        self._col_counts = defaultdict(lambda: [0] * size)
        self._diag_count = defaultdict(int)
        self._anti_diag_count = defaultdict(int)
        self._filled_cells = 0

    def at(self, row: int, col: int) -> Optional[str]:
        return self._cells[row][col]

    def is_full(self) -> bool:
        return self._filled_cells == self.size * self.size

    def _in_bounds(self, row: int, col: int) -> bool:
        return 0 <= row < self.size and 0 <= col < self.size

    def place(self, row: int, col: int, symbol: str) -> bool:
        """Places symbol at (row, col); returns True if this move wins."""
        if not self._in_bounds(row, col):
            raise OutOfBoundsError(f"({row}, {col}) is out of bounds")
        if self._cells[row][col] is not None:
            raise CellOccupiedError(f"({row}, {col}) is already occupied")

        self._cells[row][col] = symbol
        self._filled_cells += 1

        self._row_counts[symbol][row] += 1
        self._col_counts[symbol][col] += 1
        if row == col:
            self._diag_count[symbol] += 1
        if row + col == self.size - 1:
            self._anti_diag_count[symbol] += 1

        return (
            self._row_counts[symbol][row] == self.size
            or self._col_counts[symbol][col] == self.size
            or self._diag_count[symbol] == self.size
            or self._anti_diag_count[symbol] == self.size
        )

    def __str__(self) -> str:
        rows = []
        for r in range(self.size):
            rows.append(" ".join(c if c else "." for c in self._cells[r]))
        return "\n".join(rows)


class Game:
    """Orchestrates turn order, move validation and terminal-state tracking.

    No locking here: unlike the parking lot, a single tic-tac-toe game is
    played by one turn owner at a time, so there is no concurrent-writer
    scenario to guard against.
    """

    def __init__(self, size: int, players: List[Player]):
        self.board = Board(size)
        self.players = players
        self._turn = 0
        self.status = GameStatus.IN_PROGRESS
        self.winner: Optional[Player] = None

    @property
    def current_player(self) -> Player:
        return self.players[self._turn]

    def move(self, row: int, col: int) -> bool:
        if self.status != GameStatus.IN_PROGRESS:
            raise GameOverError("game is already over")

        player = self.players[self._turn]
        won = self.board.place(row, col, player.symbol)

        if won:
            self.status = GameStatus.WON
            self.winner = player
            return True
        if self.board.is_full():
            self.status = GameStatus.DRAW
            return False

        self._turn = (self._turn + 1) % len(self.players)
        return False


def _demo() -> None:
    alice = Player("Alice", "X")
    bob = Player("Bob", "O")
    game = Game(3, [alice, bob])

    moves = [(0, 0), (1, 0), (0, 1), (1, 1), (0, 2)]
    for row, col in moves:
        won = game.move(row, col)
        if won:
            print(f"{game.winner.name} wins!")
    print(game.board)


if __name__ == "__main__":
    _demo()
